core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1053,6 +1053,23 @@ private BytesReference buildMessage(long requestId, byte status, Version nodeVer
         return new CompositeBytesReference(header, messageBody, zeroCopyBuffer);
     }
 
+    protected abstract int length(Buffer buffer);
```

Comment:
I spoke to @jasontedor offline and we seem to agree on a new interface that only uses StreamInput we also going to move the length check out since it requires reading the entire message before verifying the header which can lead to system instability.